packages/@react-stately/data/src/useAsyncList.ts

```diff
@@ -249,20 +266,32 @@ export function useAsyncList<T, C = string>(options: AsyncListOptions<T, C>): As
     filterText: initialFilterText
   });
 
+  let loadingMore = useRef(false);
+
   const dispatchFetch = async (action: Action<T, C>, fn: AsyncListLoadFunction<T, C>) => {
     let abortController = new AbortController();
     try {
       dispatch({...action, abortController});
+      let previousFilterText = action.filterText ?? data.filterText;
       let response = await fn({
         items: data.items.slice(),
         selectedKeys: data.selectedKeys,
         sortDescriptor: action.sortDescriptor ?? data.sortDescriptor,
         signal: abortController.signal,
         cursor: action.type === 'loadingMore' ? data.cursor : null,
-        filterText: action.filterText ?? data.filterText
+        filterText: previousFilterText
       });
 
+      let filterText = response.filterText ?? previousFilterText;
       dispatch({type: 'success', ...response, abortController});
+      if (action.type === 'loadingMore') {
+        loadingMore.current = false;
+      }
+      // Fetch a new filtered list if filterText is updated via `load` response func rather than list.setFilterText
+      // Only do this if not aborted (e.g. user triggers another filter action before load completes)
+      if (filterText && (filterText !== previousFilterText) && !abortController.signal.aborted) {
+        dispatchFetch({type: 'filtering', filterText}, load);
+      }
```

Comment:
Need to `dispatchFetch` for the new filter text value here since a ComboBox that uses `useAsyncList` will have a controlled value and therefore won't have its `inputValue` reset automatically by the useEffect in useComboBoxState